Adopt smart pointers in FullscreenManager

csaavedra opened the pull request:
#### fab395bb04bd9f3277ad13609eb3c4ad9ccce376
<pre>
Adopt smart pointers in FullscreenManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=272235">https://bugs.webkit.org/show_bug.cgi?id=272235</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

Protect accesses to fullscreenElement() when
used non-trivially. Also fix a variable typo.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::canShowControlsManager const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::documentBackgroundColor const):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::isHiddenBehindFullscreenElement):

Canonical link: <a href="https://commits.webkit.org/277159@main">https://commits.webkit.org/277159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/616f65596a1599cf6f77661b0bde5cc99d612864

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46862 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49541 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23491 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38159 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47443 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23056 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40357 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19469 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20379 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41497 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4908 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43088 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51415 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21873 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18220 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45454 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23159 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44424 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10351 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23687 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22870 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -312,8 +312,8 @@ void FullscreenManager::cancelFullscreen()
         // This triggers finishExitFullscreen with ExitMode::Resize, which fully exits the document.
-        if (auto* fullscreenElement = topDocument->fullscreenManager().fullscreenElement())
-            topDocument->page()->chrome().client().exitFullScreenForElement(fullscreenElement);
+        if (RefPtr fullscreenElement = topDocument->fullscreenManager().protectedFullscreenElement())
```
`protectedFullscreenElement()` is unnecessary here, please keep `fullscreenElement()`. `protectedFoo()` functions are only used to avoid local variables, NOT to initialize local variables.

csaavedra replied:
But the change to use a RefPtr instead of a raw pointer is ok, right? Just that initializing from the protected version is unnecessary.

cdumez replied:
Correct.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -365,8 +365,7 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
 
-    auto element = exitingDocument->fullscreenManager().fullscreenElement();
-    if (element && !element->isConnected()) {
+    if (RefPtr element = exitingDocument->fullscreenManager().protectedFullscreenElement(); element.get() && !element->isConnected()) {
```
`protectedFullscreenElement()` is unnecessary here.

cdumez commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -113,6 +111,10 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager>, public
     WTFLogChannel& logChannel() const;
 #endif
 
+    Page* page() const { return document().page(); }
```
These can stay public. They do have use cases, even in our smart pointer world. When calling a trivial getter, using a raw pointer overload is more efficient and safe, e.g.
```
page()->protectedElement()->nonTrivialFunction();
```

if `protectedElement()` is a trivial getter, there is no reason to protect the page here, protecting the Element is enough.

cdumez commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4525,7 +4525,7 @@ Color LocalFrameView::documentBackgroundColor() const
             return { };
 
-        auto* fullscreenElement = fullscreenManager->fullscreenElement();
+        RefPtr fullscreenElement = fullscreenManager->protectedFullscreenElement();
```
No need for protectedFullscreenElement().

cdumez commented on Source/WebCore/page/ios/ContentChangeObserver.cpp:
```diff
@@ -65,7 +65,7 @@ static bool isHiddenBehindFullscreenElement(const Node& descendantCandidate)
     if (!fullscreenManager)
         return false;
-    auto* fullscreenElement = fullscreenManager->fullscreenElement();
+    RefPtr fullscreenElement = fullscreenManager->protectedFullscreenElement();
```
ditto.

cdumez reviewed (changes requested):
Please address issues.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -365,8 +365,7 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
 
-    auto element = exitingDocument->fullscreenManager().fullscreenElement();
-    if (element && !element->isConnected()) {
+    if (RefPtr element = exitingDocument->fullscreenManager().fullscreenElement(); element.get() && !element->isConnected()) {
```
You don’t need the .get() here

cdumez reviewed (approved):
LGTM with nit

csaavedra committed:
Adopt smart pointers in FullscreenManager
https://bugs.webkit.org/show_bug.cgi?id=272235

Reviewed by Chris Dumez and Ryosuke Niwa.

Protect accesses to fullscreenElement() when
used non-trivially. Also fix a variable typo.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::canShowControlsManager const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::documentBackgroundColor const):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::isHiddenBehindFullscreenElement):

Canonical link: https://commits.webkit.org/277159@main
